TwoStageRWLock for lists

Wouterdek opened the pull request:
This pull request is a smaller chunk of PR #199. It implements the new locking mechanism for lists only. See PR #199 for more details.

RolandPheasant commented:
I am in the process of adding bench marks to dd. When I am done with that I will run them against this, then assuming the is no significant performance hit will merge this PR.

RolandPheasant commented:
I have added some benchmarks to dd and compared the performance difference between the previous locking mechanism and the changes here. The good news is I see is no significant difference is performance. So let's get this in. First I will add some review notes.

RolandPheasant commented on DynamicData/Kernel/TwoStageRWLock.cs:
```diff
@@ -0,0 +1,149 @@
+	/// Before disposing this object, all locks should be released.
+	/// Failure to do so will result in a SynchronizationLockException.
+	/// </remarks>
+	public class TwoStageRWLock : IDisposable
```
This should be made internal otherwise I may end up having to support it to the public at large

RolandPheasant commented on DynamicData/Kernel/TwoStageRWLock.cs:
```diff
@@ -0,0 +1,149 @@
+﻿using System;
+using System.Collections.Generic;
```
using System.Collections.Generic;
using System.Text;

are not used, so please remove

RolandPheasant commented on DynamicData/Kernel/InternalEx.cs:
```diff
@@ -103,5 +103,12 @@ public static IDisposable ScheduleRecurringAction(this IScheduler scheduler, Fun
         }
-    }
+
+        public static void Swap<TSwap>(ref TSwap t1, ref TSwap t2)
```
Should be internal

RolandPheasant commented on DynamicData/Kernel/TwoStageRWLock.cs:
```diff
@@ -0,0 +1,149 @@
+using System.Text;
+using System.Threading;
+
+namespace DynamicData.Kernel
```
Although it does not show in this editor, can you remove tabs. This is easily achieved in vs. Select all text, Edit->Advanced->Untabbify Selected Lines. Please do the same for all the files which you have changed.

NB: I have never set up fx cop rules because I find it so intrusive and slows down editing.

Wouterdek commented:
> I have added some benchmarks to dd and compared the performance difference between the previous locking mechanism and the changes here. The good news is I see is no significant difference is performance. So let's get this in. First I will add some review notes.

Great! I have addressed your comments in my latest commit. Let me know if there is anything else that needs tweaking.

RolandPheasant commented:
Thanks. I suggest 2 more PRs. The first to add the new lock to source cache and the second to add Preview to list and cache.

I am sure what you are doing will be a great contribution to this library.
